### PR TITLE
Paginate packages for profile page in database

### PR DESCRIPTION
**Pull request by zivkan**
For #5877 and #5659 

Deploying the current dev code into an Azure App Service in the production slot, and putting my changes in a staging slot, then hitting page 2 on both sites 120 times each, App Insights tells me these results:

Version | 50th percentile | 95th percentile
---- | ----- | -----
dev code | 5,027ms | 6,150ms
these changes | 3,674ms | 4,570ms

The contents appear to be the same, except that the total package download count is different. It might be due to bad data in my database (I manually set the TotalDownloads column in the PackageRegistrations table, but didn't change the Packages table). However, I think that the value after my changes appear to be more correct than before my changes, so I'm not sure if this fixes a total downloads count bug.

The SQL that the various queries execute are horrific, but I guess that's part of using an ORM, and the performance is still better than paginating in C#. I have not tried any optimisation, I just re-used the existing query.

### CODE REVIEW

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 1]
```diff
@@ -614,15 +614,15 @@ public virtual ActionResult Profiles(string username, int page = 1)
...
-                .Where(p => p.PackageStatusKey == PackageStatus.Available)
-                .OrderByDescending(p => p.PackageRegistration.DownloadCount)
+            var packages = PackageService.GetPagedPackagesByOwner(user, page, Constants.DefaultPackageListPageSize, includeUnlisted: false)
                 .Select(p => new ListPackageItemViewModel(p, currentUser)
                 {
                     DownloadCount = p.PackageRegistration.DownloadCount
                 }).ToList();
 
-            var model = new UserProfileModel(user, currentUser, packages, page - 1, Constants.DefaultPackageListPageSize, Url);
+            var (totalPackages, totalPackageDownloadCount) = PackageService.GetTotalPackagesStatisticsForOwner(user, includeUnlisted: false);
```
Comment (loic-sharma) [inline src/NuGetGallery/Controllers/UsersController.cs]: The tuple could become annoying if we decide to add more statistics. What do you think of adding creating a new class to hold the statistics, and then passing that class to the `UserProfileModel`?

File: src/NuGetGallery/Services/PackageService.cs [hunk 2]
```diff
@@ -226,9 +226,27 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
         }
 
-        public IEnumerable<Package> FindPackagesByOwner(User user, bool includeUnlisted, bool includeVersions = false)
+        public IEnumerable<Package> GetPagedPackagesByOwner(User user, int page, int pageSize, bool includeUnlisted, bool includeVersions = false)
         {
-            return GetPackagesForOwners(new[] { user.Key }, includeUnlisted, includeVersions);
+            return GetPackagesForOwners(new[] { user.Key }, includeUnlisted, includeVersions)
+                .Where(p=>p.PackageStatusKey == PackageStatus.Available)
+                .OrderByDescending(p=>p.PackageRegistration.DownloadCount)
+                .Skip((page-1)*pageSize)
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageService.cs]: >page-1 [](start = 23, length = 6)

I know that the profile page is 1-based indexed, but I think that people would expect this to be 0-based indexed. I would either:

1. Make this method be 0-based indexed, and update the UsersController
2. Document that this method is 1-based indexed in the `IPackageService`

File: src/NuGetGallery/Services/PackageService.cs [hunk 3]
```diff
@@ -226,9 +226,27 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
             return package;
         }
 
-        public IEnumerable<Package> FindPackagesByOwner(User user, bool includeUnlisted, bool includeVersions = false)
+        public IEnumerable<Package> GetPagedPackagesByOwner(User user, int page, int pageSize, bool includeUnlisted, bool includeVersions = false)
         {
-            return GetPackagesForOwners(new[] { user.Key }, includeUnlisted, includeVersions);
+            return GetPackagesForOwners(new[] { user.Key }, includeUnlisted, includeVersions)
+                .Where(p=>p.PackageStatusKey == PackageStatus.Available)
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageService.cs]: >p=>p [](start = 23, length = 4)

Missing spaces on this line and the next: `p=>p` should be `p => p`

File: src/NuGetGallery/Services/IPackageService.cs [hunk 4]
```diff
@@ -29,7 +29,9 @@ public interface IPackageService : ICorePackageService
 
         Package FindAbsoluteLatestPackageById(string id, int? semVerLevelKey);
 
-        IEnumerable<Package> FindPackagesByOwner(User user, bool includeUnlisted, bool includeVersions = false);
+        IEnumerable<Package> GetPagedPackagesByOwner(User user, int page, int pageSize, bool includeUnlisted, bool includeVersions = false);
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageService.cs]: >bool includeVersions = false [](start = 110, length = 28)

What uses `includeVersions`? It seems that this parameter can be removed from both `GetPagedPackagesByOwner` and `GetTotalPackagesStatisticsForOwner`

File: src/NuGetGallery/Services/PackageService.cs (hunk 2 above)
Comment (agr) [reply src/NuGetGallery/Services/PackageService.cs]: I would suggest passing the skip value into this method instead. I don't think calculating the skip belongs here.

File: src/NuGetGallery/ViewModels/UserProfileModel.cs [hunk 5]
```diff
@@ -8,43 +8,33 @@ namespace NuGetGallery
 {
     public class UserProfileModel
     {
-        public UserProfileModel(User user, User currentUser, List<ListPackageItemViewModel> allPackages, int pageIndex, int pageSize, UrlHelper url)
+        public UserProfileModel(User user, User currentUser, List<ListPackageItemViewModel> pagedPackages, int totalPackages, long totalPackageDownloadCount, int pageIndex, int pageSize, UrlHelper url)
```
Comment (agr) [inline src/NuGetGallery/ViewModels/UserProfileModel.cs]: Time to split argument list into individual lines (one line per argument), I think.

File: tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs [hunk 6]
```diff
@@ -987,8 +987,11 @@ public void ReturnsTheMostRecentVersionWhenSemVerLevel2()
             }
         }
 
-        public class TheFindPackagesByOwnerMethod : TheFindPackagesByOwnersMethodsBase
+        public class TheGetPagedPackagesByOwnerMethod : TheFindPackagesByOwnersMethodsBase
         {
+            private int page = 1;
```
Comment (agr) [inline tests/NuGetGallery.Facts/Services/PackageServiceFacts.cs]: `const`?

File: src/NuGetGallery/Services/PackageService.cs [hunk 7]
```diff
@@ -226,9 +226,27 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
+
+        public (int totalPackages, long totalPackageDownloadCount) GetTotalPackagesStatisticsForOwner(User user, bool includeUnlisted, bool includeVersions = false)
+        {
+            var packages = GetPackagesForOwners(new[] { user.Key }, includeUnlisted, includeVersions);
+            var result = packages.Select(p => new
+            {
+                TotalPackages = packages.Count(),
+                TotalPackageDownloadCount = packages.Sum(p2 => (long)p2.PackageRegistration.DownloadCount)
+            })
+                .First();
```
Comment (agr) [inline src/NuGetGallery/Services/PackageService.cs]: Inconsistent indentation.

Comment (agr) [reply src/NuGetGallery/Services/PackageService.cs]: I'd prefer something like

```
            var result = packages
                .Select(p => new
                    {
                        TotalPackages = packages.Count(),
                        TotalPackageDownloadCount = packages.Sum(p2 => (long)p2.PackageRegistration.DownloadCount)
                    })
                .First();
```

(not sure about indenting the curly block, but this looks right to me: it's inner block for `Select`, so should be indented further)

**Comment by skofman1** [comment]
@zivkan , can you provide a link to your webapp so we can try things out?

File: src/NuGetGallery/Services/PackageService.cs (hunk 7 above)
Comment (zivkan) [reply src/NuGetGallery/Services/PackageService.cs]: I also prefer your suggested style, but I don't have resharper installed, and I couldn't find a .editorconfig/VS formatting option to make the auto-formatting do that. I prefer using auto-formatting styles, so if someone reformats the entire document, it doesn't cause changes to code they didn't intend to change.

**Comment by zivkan** [comment]
My web app URLS are:
https://anzivk-nuget.azurewebsites.net/profiles/microsoft
https://anzivk-nuget-staging.azurewebsites.net/profiles/microsoft

the times I listed in the PR message are for the microsoft account with 3,760 packages, just like in prod (although my database doesn't have other package registrations, and only has 1 package version per registration).

Sorting is currently broken too, so I have some more work to do.

**Comment by loic-sharma** [comment]
Adding @xavierdecoster since he will know the most about this.

That interceptor gets the latest downloads within 3 hours (it uses the `downloads.v1.json` auxiliary file, which is refreshed by the [CreateAzureCdnWarehouseReports job](https://github.com/NuGet/NuGet.Jobs/blob/master/src/Stats.CreateAzureCdnWarehouseReports/Job.cs) every 3 hours. So, it seems that the interceptor's purpose is to get fresher download counts than what's in the Gallery DB. If that's true, I think we can skip the interceptor entirely. What do you think @xavierdecoster

**Comment by xavierdecoster** [comment]
The main reason for this interception is to be consistent across the gallery in the numbers we show (on the web site, in v3 and v2 protocols). Due to the time it takes to generate the blobs and aggregate the download counts into the galllery database, there may be a time window during which results would be inconsistent.

If newer data is available in the blobs, v2 should also return these and discard the not-yet-updated counts in the db. In other words: download counts from blob storage overrule those found in the gallery database.

**Comment by xavierdecoster** [comment]
I remember some people had strong opinions on data (in)consistency, which is why this interceptor was built in the first place.

What if we decided to no longer show the download counts on this particular view? 
This way, stats have nothing to do with rendering this view, and it could also help solve another issue: https://github.com/NuGet/NuGetGallery/issues/5909

**Comment by skofman1** [comment]
Sounds like we can't easily get a consistent experience. Will return this issue to the backlog since the perf gain is not sufficient to justify the experience impact.

**Pull request closed without merging**